### ISSUE TITLE
Publish SNAPSHOTs on Sonatype

### ISSUE THREAD

**Issue by sschuberth** [uwolfer/gerrit-rest-java-client#34]
Can we get SNAPSHOTs published on Sonatype? We'd esp. need `gerrit-rest-java-client-0.8.8-SNAPSHOT.jar` for @EdwinKempin's [ReviewIT](https://gerrit-review.googlesource.com/#/projects/apps/reviewit,dashboards/default) Android app :-)

**Comment by EdwinKempin** [comment]
This request is about the httpclient-android branch.

**Comment by sschuberth** [comment]
> This request is about the httpclient-android branch.

Oh, sorry, I wasn't aware of that. In that case, maybe it's too confusing to publish a SNAPSHOT that is not built from master? What's the plan to integrate httpclient-android to master?

**Comment by uwolfer** [comment]
Great work on ReviewIT! :+1: 

I have not looked into publishing snapshots yet, but I release "normal" releases to Sonatype (but not the httpclient-branch yet). If you know any easy (probably almost automated way) to publish snapshots to Sonatype, please let me know.

I really would like to have just one branch, but this is not possible because of the [httpclient issue on Android](https://hc.apache.org/httpcomponents-client-4.3.x/android-port.html)... (at least the last time I have looked into it together with the author of mGerrit, which was probably almost a year ago). If there is a good way to implement it the same way for Android and "normal" JRE, it would be great. If you have any idea here, please let me know.

**Comment by sschuberth** [comment]
The httpclient issue is indeed annoying. @uwolfer just wondering, have you considered instead using using [Retrofit](https://github.com/square/retrofit)? Being based on [OkHttp](https://github.com/square/okhttp) it should work equally well on Android and JVM.

**Comment by uwolfer** [comment]
mGerrit developer @p4r4n01d has tried to port it to OkHttp some time ago but he decided to stay with HttpClient after some time because of missing features.

His work-in-progress (abandoned) can be found here: https://github.com/p4r4n01d/gerrit-rest-java-client/commits/okhttp

I have not looked into Retrofit yet.

**Comment by sschuberth** [comment]
@p4r4n01d Would you have any details on the missing features and whether using [Retrofit](https://square.github.io/retrofit/) on top of OkHttp would help?

**Comment by sschuberth** [comment]
@uwolfer WRT publishing snapshots to Sonatype, have you checked the videos at http://central.sonatype.org/pages/ossrh-guide.html? For the *httpclient-android* branch you could probably publish with the classifier set to `android`, and publish this artifact in parallel to the one from the *master* branch.

**Comment by p4r4n01d** [comment]
@sschuberth I have done some searching through my emails from when I was working on this and from what I remember, only the Apache library supports all the methods of authentication that Gerrit (and this library) uses. I had a lot of problems with digest authentication using OkHttp as they didn't properly support it. This is a big problem if Gerrit's default authentication method is digest as I think it is.

I would be really happy porting this to use OkHttp, but it is not an option if half the features in this library are dropped as OkHttp doesn't support them. I would also be happy if the Guava dependency is removed (that is a big library which adds a TON of methods) but that is another story...

I sent this to @uwolfer on 30/05:
> For OkHttp: I have essentially given up trying to do digest authentication without apache. I tried to match the response header as closely to the included apache implementation, but I always get 403 forbidden back when the apache one succeeds (I was using a real server for testing). As it can authenticate and fetch/parse the response, the next step is to attempt the migration for mGerrit, then I'll worry about things like preemptive auth.

**Comment by p4r4n01d** [comment]
@sschuberth Where can I find out more about this ReviewIT app? I am pretty interested in other code review apps. Is there anywhere I can have a look at other than Gerrit code review (which is rather ironic btw)?

**Comment by sschuberth** [comment]
@p4r4n01d The ReviewIt app currently is only hosted on [googlesource.com](https://gerrit-review.googlesource.com/#/admin/projects/apps/reviewit), but you can anonymously clone it via:

    git clone https://gerrit.googlesource.com/apps/reviewit

so that should not be a big issue. @EdwinKempin's original announcement in the repo-discuss group is [over here](https://groups.google.com/d/msg/repo-discuss/VaSJGTPUpd4/C3HbU2lSCgAJ).

**Comment by sschuberth** [comment]
> I would also be happy if the Guava dependency is removed (that is a big library which adds a TON of methods) but that is another story...

Is method count on Android still an issue for you in multi-dex times?

**Comment by p4r4n01d** [comment]
Thanks for that link.

It is not really an issue.

**Comment by sschuberth** [comment]
> I had a lot of problems with digest authentication using OkHttp as they didn't properly support it.

They indeed still do not seem to support it: https://github.com/square/okhttp/issues/205

However, the https://github.com/rburgst/okhttp-digest library might come to the rescue, though I haven't tried it yet.

**Comment by sschuberth** [comment]
Closing the request for a SNAPSHOT published from the `httpclient-android` branch in favor of using [JitPack](https://jitpack.io/) as suggested [here](https://gerrit-review.googlesource.com/75995).

**Comment by uwolfer** [comment]
@sschuberth: Nice solution! Probably other users of the Android branch can also use it.
